Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,892 @@
+class Functor m => Monad m where
+  return :: a -> m a
+  (>>=) :: m a -> (a -> m b) -> m b
+```
+
+This definition needs to additionally satisfy the monad laws:
+
+```haskell
+(m >>= g) >>= h = m >>= (\x -> g x >>= h) -- associativity
+return a >>= h = h a -- left identity
+m >>= return = m -- right identity
+```
+
+If you look at the monad laws and squint a bit (or perhaps a lot), you might notice that these laws seem somewhat similar to the monoid laws.
+The names of the laws are the same, at least.
+Our choice of the definition makes it a little harder to see the connection, though.
+If we instead rewrite these laws using the Kleisli composition
+
+```haskell
+(>=>) :: Monad m => (a -> m b) -> (b -> m c) -> (a -> m c)
+f >=> g = \x -> f x >>= g
+```
+
+it becomes much more apparent[^equivalence]:
+
+[^equivalence]: The proof of equivalence is left as an exercise to the reader.
+    Here, by equivalence we mean that the monad laws expressed via Kleisli arrows imply the monad laws expressed via `>>=` and vice versa.
+    The proof itself is a straightforward application of equational reasoning.
+    It's easier to start from the Kleisli arrow form.
+
+```haskell
+(f >=> g) >=> h = f >=> (g >=> h) -- associativity
+return >=> f = f -- left identity
+f >=> return = f -- right identity
+```
+
+That is why a monad is indeed a monoid: it satisfies the same laws, and the "values" are of the type `Functor f => a -> f b`.
+Note, though, that this is not the category of endofunctors.
+However, there exists a natural transformation to the category of endofunctors[^endofunctors].
+
+[^endofunctors]: One could alternatively define monads via `join :: m (m a) -> m a` instead of `>>=`, which would make it a little more evident that monads are monoids over endofunctors, but to explain this properly, we would need to use a bit more of category theory jargon.
+    There is an excellent [r/math post][endofunctors] on the subject.
+
+    [endofunctors]: https://www.reddit.com/r/math/comments/ap25mr/a_monad_is_a_monoid_in_the_category_of/
+
+We can guess a few things, based on the intuition we gained looking at free monoids.
+First, we might note that `return` and `>=>` correspond to `mempty` and `(<>)`, respectively.
+Second, we can expect that for any `Functor` `f :: Type -> Type`, there is a corresponding free `Monad` `Free f :: Type -> Type`, the same as for free monoids.
+
+## Free monads
+
+Let us first try to construct `Free f` by analogy.
+We've seen that lists are free monoids.
+The typical definition of a list type in Haskell looks something like this:
+
+```haskell
+data List a = Nil | Cons a (List a)
+```
+
+Now, `List a` is a type, but `Free f` is a unary type constructor, so we'd need one more type argument.
+Doing that directly yields
+
+```haskell
+data Free f a = Nil a | Cons f (Free f) a
+```
+
+We can't define such a data type, due to `f` not being a plain type and hence can't be an argument of a data constructor, but we're actually very close.
```

Comment:
```suggestion
We can't define such a data type due to `f` not being a plain type (hence, it can't be an argument of a data constructor), but we're actually very close.
```

Or you can rephrase in another way, but: no comma after due and there is something odd in the original sentence around hence.